app/controllers/dfe_sign_in_controller.rb

```diff
@@ -2,9 +2,12 @@ class DfESignInController < ActionController::Base
   protect_from_forgery except: :bypass_callback
 
   def callback
+    @target_path = session['post_dfe_sign_in_path']
+
+    reset_session # prevents session fixation attacks and impersonation bugs
```

Comment:
I've had a look at other uses of `session` in the app:

```
app/lib/wizard_state_stores/session_store.rb:      @session[@key] = value
app/helpers/view_helper.rb:    if session[:confirmed_environment_at] && session[:confirmed_environment_at] > 5.minutes.ago
app/controllers/concerns/cookie_preference_concerns.rb:    session[:display_cookie_consent_confirmation] = true
app/controllers/candidate_interface/content_controller.rb:      session[:previous_referer] = request.referer
app/controllers/provider_interface/content_controller.rb:      session[:previous_referer] = request.referer
app/controllers/support_interface/sessions_controller.rb:        session[:confirmed_environment_at] = Time.zone.now
```

Devise uses `session`, so using `reset_session` destroys any candidate sessions if you log in as a provider or support user (the callback mechanism is the same for both). Shouldn't be a problem on production, as re-authenticating will kill both provider and candidate impersonations, which is more consistent.

We only have one wizard that uses `SessionStore`, the 'reconfirm deferred offer' one. Losing progress for this one shouldn't be that much of an issue. When we revisit, we'll probably move this to `RedisStore`, anyway.

I am keen to use `reset_session` because it is considered best practice. Maybe we should take the approach of obliterating everything and only preserving what we need? For example, we could preserve `session[:display_cookie_consent_confirmation]` and `session[:previous_referer]`.

Reply:
Discussed elsewhere, I've updated the code to preserve all existing values, but:
- change the `session_id` (as a countermeasure for session fixation)
- drop any previous provider impersonation (to prevent impersonation bugs)